sub/sd_ass.c

```diff
@@ -448,6 +448,32 @@ static void configure_ass(struct sd *sd, struct mp_osd_res *dim,
     if (converted)
         ass_track_set_feature(track, ASS_FEATURE_WRAP_UNICODE, 1);
 #endif
+    if (converted) {
+        bool override_playres = true;
+        char **ass_force_style_list = opts->ass_force_style_list;
+        for (int i = 0; ass_force_style_list && ass_force_style_list[i]; i++) {
+            if (bstr_find0(bstr0(ass_force_style_list[i]), "PlayResX") >= 0)
+                override_playres = false;
+        }
+
+        // srt ass from ffmpeg has fixed PlayerResX of 384 (and aspect of 4:3).
+        // starting with libass f08f8ea5 (pre 0.17) PlayerResX affects shadow
+        // and border widths, among others, so to render borders and shadows
+        // correctly, we adjust PlayerResX according to the video aspect.
+        // But PlayerResX also affects margins, so we adjust those too.
+        // This should ensure basic srt-to-ass ffmpeg conversion has correct
+        // borders, but there could be other issues with some srt extensions
+        // and/or different source formats which would be exposed over time.
+        // we make these adjustments only if the user didn't set PlaterResX.
+        if (override_playres) {
+            int vidw = dim->w - (dim->ml + dim->mr);
+            int vidh = dim->h - (dim->mt + dim->mb);
+            track->PlayResX = track->PlayResY * (double)vidw / MPMAX(vidh, 1);
+            double fix_margins = track->PlayResX / 384.0;
+            track->styles->MarginL = round(track->styles->MarginL * fix_margins);
+            track->styles->MarginR = track->styles->MarginL;
```

Comment:
The general idea of overriding `PlayResX` and Margins seems good, but the specific implementation can be improved. As for the calculation of `vidw` and `vidh` itself I’ll just assume it is correct, but can't actually verify.
 - if `vidw` or `vidh` can be zero, it would be better to cap *both* to `>= 1` and move the `MPMAX` to their respective definitions.
 - hardcoding `384` seems brittle and may exacerbate problems with formats whose ffmpeg converter uses custom `PlayRes*`. Instead save the old `PlayResX` value to a temporary variable before overriding it.
 - i don't see why the fixup should enforce `MarginL == MarginR`. Scaling both independently seems better.
 
 ```diff
diff --git a/sub/sd_ass.c b/sub/sd_ass.c
index d3612f5514..32ea41991f 100644
--- a/sub/sd_ass.c
+++ b/sub/sd_ass.c
@@ -464,12 +464,13 @@ static void configure_ass(struct sd *sd, struct mp_osd_res *dim,
         // Do NOT change anything if an user override already set PlayResX.
         // TODO: also exclude converters with custom PlayRes?
         if (override_playres) {
-            int vidw = dim->w - (dim->ml + dim->mr);
-            int vidh = dim->h - (dim->mt + dim->mb);
-            track->PlayResX = track->PlayResY * (double)vidw / MPMAX(vidh, 1);
-            double fix_margins = track->PlayResX / 384.0;
+            int vidw = MPMAX(dim->w - (dim->ml + dim->mr), 1);
+            int vidh = MPMAX(dim->h - (dim->mt + dim->mb), 1);
+            double old_prx = track->PlayResX;
+            track->PlayResX = track->PlayResY * vidw / (double) vidh;
+            double fix_margins = track->PlayResX / old_prx;
             track->styles->MarginL = round(track->styles->MarginL * fix_margins);
-            track->styles->MarginR = track->styles->MarginL;
+            track->styles->MarginR = round(track->styles->MarginR * fix_margins);
         }
     }
 }
 ```

Reply:
I don't think vidw or vidh can be zero, a lot of other things would blow up before this piece of code would even be reached if that were the case. @avih suggested that I wrap it in MPMAX.

288x384 is hardcoded everywhere in sub/*, I eventually intend to clean this up but for now I'll save old PlayResX.

MarginL == MarginR is something that we already do, so I simply copied the behavior. https://github.com/mpv-player/mpv/blob/master/sub/ass_mp.c#L76-L77

Reply:
> @avih suggested that I wrap it in MPMAX.

Yeah, we divide by `vidh`, so it can't be 0. It's guarded just like earlier in this function. If `vidw` is 0 then whatever, but it won't blow due to division by zero.

Reply:
> If vidw is 0 then whatever, but it won't blow due to division by zero.

`PlayRes* <= 0` is not a valid value, so it will be treated as unset(or invalidly set) and ASS’ mechanism for deriving a default kicks in. While in most cases I’d expect this to still lead to `PlayResX = 1` in the end, this may not hold true for all edge cases and it is not a particular clean approach.
Just clamping `vidw`/`widh` as I previously suggested doesn't fully solve it either though. The following would probably be best:
```C
track->PlayResX = MPMAX(track->PlayResY * vidw / (double) MPMAX(vidh, 1), 1);
```

> MarginL == MarginR is something that we already do

I’m not that familiar with this part of mpv, but I’m guessing this just sets the initial default (same with hardcoded `288×384`). Otherwise e.g. user overrides like `MarginL=5,MarginR=50` wouldn't work. But we want them to work and then scale them afterwards.

In either case I personally think it is best to not hardcode any values or assumptions about how the Style was preinitialised at this part. This will only cause additional hassle should those other areas ever change. In the end this up to mpv devs to decide though.

Reply:
> `PlayRes* <= 0` is not a valid value, so it will be treated as unset(or invalidly set) and ASS’ mechanism for deriving a default kicks in.

That's absolutely fine as far as I'm concerned. If the width is 0 then nothing sensible can result either way. The goal here is only to avoid division by zero.

> Otherwise e.g. user overrides like `MarginL=5,MarginR=50` wouldn't work

This is already fixed at the updated PR code from yesterday. Thanks for noticing it in the first palce.

> In either case I personally think it is best to not hardcode any values

If you refer to the 384, then we already added a comment why it's fixed like that. This should be enough until we maybe (or never) unify all the 384 values into a single constant.

Also, I'd appreciate if you could reply to my question about the "misleading phrasing". Thanks.

Reply:
>  refer to the 384, then we already added a comment why it's fixed like that. This should be enough until we maybe (or never) unify all the 384 values into a single constant.

this new comment
```
ffmpeg and mpv use a default PlayResX of 384 when it is not known, this comes from VSFilter
```
 does not explain why you are expecting hardcoding this here to be a good idea for e.g. the aforementioned ffmpeg-converted sub with custom PlayRes or why you think ffmpeg won't ever change their default `PlayRes` sizes.
“Comes from VSFilter” is no justification at this place. Converters do not need to care what defaults ASS uses for unset `PlayRes` since they can just set whatever valid values they want. This entails being free to chose their own defaults whenever and to whatever valid value they like. And ASS defaults behaves differently depending on if both or only one `PlayRes*` value is unknown and what the known value is, so it isn't even always `384`.

I continue to believe this is a bad idea, but if you’re determined to stick with this then do so.

Reply:
> `PlayRes* <= 0` is not a valid value, so it will be treated as unset(or invalidly set) and ASS’ mechanism for deriving a default kicks in.

That's absolutely fine as far as I'm concerned. If the width is 0 then nothing sensible can result either way. The goal here is only to avoid division by zero.

> Otherwise e.g. user overrides like `MarginL=5,MarginR=50` wouldn't work

This is already fixed at the updated PR code from yesterday. Thanks for noticing it in the first palce.

> In either case I personally think it is best to not hardcode any values

If you refer to the 384, then we already added a comment why it's fixed like that. This should be enough until we maybe (or never) unify all the 384 values into a single constant.

Also, I'd appreciate if you could reply to my question about the "misleading phrasing". Thanks.

> I continue to believe this is a bad idea, but if you’re determined to stick with this then do so.

What exactly is a bad idea here? using 384 with a comment instead of using it as a constant or some `#define` ?

Reply:
> I continue to believe this is a bad idea, but if you’re determined to stick with this then do so.

What exactly is a bad idea here? using 384 with a comment instead of using it as a constant or some `#define` ?

I don't think I understand what you think would be a better idea...

Reply:
> What exactly is a bad idea here? using 384 with a comment instead of using it as a constant or some `#define` ?

Using a constant variable or `#define` would be no better. Assuming `PlayResX` always were `384` before is the problem and I already pointed out cases for which this assumption doesn't hold.
The better idea is exactly what I wrote in my initial code comment and the accompaniying suggested diff:
> - hardcoding 384 seems brittle wrt to future ffmpeg changes and may exacerbate problems with formats whose ffmpeg converter uses custom PlayRes*. Instead save the old PlayResX value to a temporary variable before overriding it.
>
> [...]
> `double fix_margins = track->PlayResX / old_prx;`